File: Two Pointers/125. Valid Palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        n = len(s)
        lo = 0
        hi = n - 1
        
        while lo <= hi:
            if not s[lo].isalnum():
                lo += 1
                continue

            if not s[hi].isalnum():
                hi -= 1
                continue
            
            if s[lo].lower() != s[hi].lower():
                return False
            
            lo += 1
            hi -= 1
        
        return True

# Time: O(n)
# Space: O(1)

File: Two Pointers/125. Valid Palindrome/125. Valid Palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function (s) {
  let l = 0;
  let r = s.length - 1;
  while (l < r) {
    if (!/^[a-z0-9]$/i.test(s[l])) {
      l++;
      continue;
    }

    if (!/^[a-z0-9]$/i.test(s[r])) {
      r--;
      continue;
    }

    if (s[l].toLowerCase() !== s[r].toLowerCase()) return false;
    l++;
    r--;
  }
  return true;
};

// Time: O(n)
// Space: O(1)
